src/prefect/flows.py

```diff
@@ -612,11 +612,15 @@ def serialize_parameters(self, parameters: Dict[str, Any]) -> Dict[str, Any]:
             if self.ismethod and value is self.fn.__prefect_self__:
                 continue
             try:
+                if isinstance(value, (PrefectFuture, State)):
+                    # Don't call jsonable_encoder() on a PrefectFuture or State to
+                    # avoid triggering a __getitem__ call
+                    raise TypeError
```

Comment:
I'd rather not raise an exception to trigger the right behavior and instead make it explicit; so maybe in line 614 above it can be:
```python
elif isinstance(value, (PrefectFuture, State)):
    serialized_parameters[key] = f"<{type(value).__name__}>"
```

Reply:
> small request to move the logic out of the try / except otherwise I think this is reasonable to support 

@cicdw: Sure, that is understandable. I went with an `if` and then `continue` here because an `elif` will still cause the `jsonable_encoder` to be called. Please take a look and let me know if this is more in line with your style suggestion.

Thank you for the review!

> I'd love to better understand your use case for keying into Prefect futures too! (I assume it's to build an async / deferred graph that doesn't force-resolve the future when you attempt to key into it? Is there a write-up or issue you could share where you explain more?)

That is correct! I am writing a library of scientific workflows that run using Prefect, and the results of the `@tasks` are typically dictionaries with a variety of keys and values, say something like the following toy example:

```python
{"input_params": {"a": 1, "b": 2}, "results": {"energy": 1.0, "converged": True}`
```

If the `@task`s don't take the full dictionary as input but rather one component of it (typically, in my case, a part of the results from a prior `@task`), then I need to index the `PrefectFuture` but don't want to introduce a blocking `.result()` or `.wait()` call. I want to defer the call to the indexing, and that is the high-level reason why this PR is relevant. Hopefully I can have some time to put together a short blog on how we're using Prefect for materials science simulations and highlight a few of these kinds of gotchas that may not be as relevant for the data engineering community.

Reply:
All of that makes sense, thank you!